lib/mix/lib/mix/dep.ex

```diff
@@ -78,71 +78,115 @@ defmodule Mix.Dep do
   @doc """
   Returns loaded dependencies from the cache for the current environment.
 
+  If dependencies have not been cached yet, they are loaded
+  and then cached.
+
   Because the dependencies are cached during deps.loadpaths,
   their status may be outdated (for example, `:compile` did not
   yet become `:ok`). Therefore it is recommended to not rely
   on their status, also given they haven't been checked
   against the lock.
-
-  If MIX_NO_DEPS is set, we return an empty list of dependencies
-  without loading them.
   """
   def cached() do
-    cond do
-      System.get_env("MIX_NO_DEPS") in ~w(1 true) ->
-        []
+    if project = Mix.Project.get() do
+      read_cached_deps(project, Mix.env()) || load_and_cache()
+    else
+      load_and_cache()
+    end
+  end
+
+  @doc """
+  Returns loaded dependencies recursively and caches it.
+
+  The result is cached for future `cached/0` calls.
+
+  ## Exceptions
+
+  This function raises an exception if any of the dependencies
+  provided in the project are in the wrong format.
+  """
+  def load_and_cache() do
+    env = Mix.env()
+
+    case Mix.ProjectStack.top_and_bottom() do
+      {%{name: top, config: config}, %{name: bottom}} ->
+        write_cached_deps(top, env, load_and_cache(config, top, bottom, env))
+
+      _ ->
+        converge(env: env)
+    end
+  end
 
-      project = Mix.Project.get() ->
-        key = {:cached_deps, Mix.env(), project}
+  defp load_and_cache(_config, top, top, env) do
+    converge(env: env)
+  end
 
-        Mix.ProjectStack.read_cache(key) ||
-          Mix.ProjectStack.write_cache(key, loaded(env: Mix.env()))
+  defp load_and_cache(config, _top, bottom, _env) do
+    {_, deps} = Mix.ProjectStack.read_cache({:cached_deps, bottom})
+    apps = [Keyword.fetch!(config, :app)]
+    get_deps(deps, tl(Enum.uniq(get_children(deps, apps, apps))))
+  end
 
-      true ->
-        loaded(env: Mix.env())
+  defp read_cached_deps(project, env) do
+    case Mix.ProjectStack.read_cache({:cached_deps, project}) do
+      {^env, deps} -> deps
+      _ -> nil
```

Comment:
Maybe we should include the env in the cache key instead. I think this can be hit with aliases or things like `mix do compile, test`.